utilities/vmware_guest_operations.py

```
@@ -247,3 +249,80 @@ def detect_vmware_ip_origins_via_guest_ops(
+    """
+    if not all(c.isalnum() or c in "-_." for c in marker_content):
+        raise ValueError(f"Marker content contains shell-unsafe characters: {marker_content!r}")
+
+    try:
+        guest_username = source_provider_data["guest_vm_linux_user"]
+        guest_password = source_provider_data["guest_vm_linux_password"]
+    except KeyError as e:
+        raise ValueError(
+            f"Linux VM credentials not found in provider config: {e}. "
+            "Required: guest_vm_linux_user, guest_vm_linux_password"
+        ) from e
```

Ruff (0.15.9)

[warning] Avoid specifying long messages outside the exception class

(TRY003)